osquery/utils/fgets_buffer.h

```diff
@@ -0,0 +1,201 @@
+/**
+ *  Copyright (c) 2019-present, osquery Foundation
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string.h>
+#include <string>
+#include <vector>
+
+#include <osquery/filesystem/fileops.h>
+
+/*
+ * This all exists to ensure that we don't get hangs when
+ * using fgets() on pipes.  See issue #4810 for details.
+ */
+
+/**
+ * @brief Abstraction of a non blocking file for use with FgetsBuffer
+ */
+struct NonblockingFile {
+  virtual bool isValid() = 0;
+
+  virtual bool isDataAvail() = 0;
+
+  virtual int64_t read(std::vector<char>& buf) = 0;
+
+  virtual void close() = 0;
+
+  virtual ~NonblockingFile() {}
+};
+
+typedef std::unique_ptr<NonblockingFile> SPNonblockingFile;
+
+#ifdef OSQUERY_POSIX
+
+/**
+ * Implementation of non-blocking file access.
+ */
+class NonblockingFileImpl : public NonblockingFile {
+ public:
+  NonblockingFileImpl(std::string filepath,
+                      uint32_t selectTimeoutUsec = 5000000)
+      : NonblockingFile(),
+        path_(filepath),
+        selectTimeoutUsec_(selectTimeoutUsec) {
+    fd_ = ::open(path_.c_str(), O_RDONLY | O_NONBLOCK);
+  }
+
+  virtual ~NonblockingFileImpl() {}
```

Comment:
Shouldn't this attempt to close open FDs?

Reply:
nice catch